core/event.cpp

```diff
@@ -93,5 +96,133 @@ Variable broadcast_weights(const DataArrayConstView &events) {
+      overloaded{[](scipp::index &c, const auto &list) { c = list.size(); },
+                 transform_flags::expect_no_variance_arg<0>});
+  return sizes;
+}
+
+/// Reserve memory in all event lists in `events`, based on `capacity`.
+///
+/// To avoid pessimizing reserves, this does nothing if the new capacity is less
+/// than the typical logarithmic growth. This yields a 5x speedup in some cases,
+/// without apparent negative effect on the other cases.
+void reserve(const VariableView &events, const VariableConstView &capacity) {
+  transform_in_place<
+      pair_custom_t<std::pair<event_list<double>, scipp::index>>,
+      pair_custom_t<std::pair<event_list<float>, scipp::index>>,
+      pair_custom_t<std::pair<event_list<int64_t>, scipp::index>>,
+      pair_custom_t<std::pair<event_list<int32_t>, scipp::index>>>(
+      events, capacity,
+      overloaded{[](auto &&sparse_, const scipp::index capacity_) {
+                   if (capacity_ > 2 * scipp::size(sparse_))
+                     return sparse_.reserve(capacity_);
+                 },
+                 transform_flags::expect_no_variance_arg<1>,
+                 [](const units::Unit &, const units::Unit &) {}});
+}
+
+namespace filter_detail {
+template <class T>
+using make_select_args = std::tuple<event_list<T>, span<const T>>;
+template <class T, class Index>
+using copy_if_args = std::tuple<event_list<T>, event_list<Index>>;
+
+constexpr auto copy_if = [](const VariableConstView &var,
+                            const VariableConstView &select) {
+  return transform<std::tuple<
+      copy_if_args<double, int32_t>, copy_if_args<float, int32_t>,
+      copy_if_args<int64_t, int32_t>, copy_if_args<int32_t, int32_t>,
+      copy_if_args<double, int64_t>, copy_if_args<float, int64_t>,
+      copy_if_args<int64_t, int64_t>, copy_if_args<int32_t, int64_t>>>(
+      var, select,
+      overloaded{
+          transform_flags::expect_no_variance_arg<1>,
+          [](const auto &var_, const auto &select_) {
+            using VarT = std::decay_t<decltype(var_)>;
+            using Events = event_list<typename VarT::value_type>;
+            const auto size = scipp::size(select_);
+            if constexpr (detail::is_ValuesAndVariances_v<VarT>) {
+              std::pair<Events, Events> out;
+              out.first.reserve(size);
+              out.second.reserve(size);
+              for (const auto i : select_) {
+                out.first.push_back(var_.values[i]);
+                out.second.push_back(var_.variances[i]);
+              }
+              return out;
+            } else {
+              Events out;
+              out.reserve(size);
+              for (const auto i : select_)
+                out.push_back(var_[i]);
+              return out;
+            }
+          },
+          [](const units::Unit &var_, const units::Unit &) { return var_; }});
+};
+
+template <class T>
+const auto make_select = [](const DataArrayConstView &array, const Dim dim,
```

Comment:
Looks good and logic makes sense. I think a few comments explaining the purpose of this function would be useful (same for others here).

Reply:
Added some doxygen 👍